sdks/java/io/jdbc/src/test/java/org/apache/beam/sdk/io/jdbc/JdbcIOIT.java

```diff
@@ -0,0 +1,190 @@
+   *
+   * <p>Note that IT read tests must not do any data table manipulation (setup/clean up.)
+   * @throws SQLException
+   */
+  @Test
+  public void testRead() throws SQLException {
+    DataSource dataSource = getDataSource();
+
+    String tableName = JdbcTestDataSet.READ_TABLE_NAME;
+
+    TestPipeline pipeline = TestPipeline.create();
+    //Pipeline pipeline = Pipeline.create(TestPipeline.testingPipelineOptions());
+
+    PCollection<KV<String, Integer>> output = pipeline.apply(JdbcIO.<KV<String, Integer>>read()
+            .withDataSourceConfiguration(JdbcIO.DataSourceConfiguration.create(dataSource))
+            .withQuery("select name,id from " + tableName)
+            .withRowMapper(new CreateKVOfNameAndId())
+            .withCoder(KvCoder.of(StringUtf8Coder.of(), BigEndianIntegerCoder.of())));
+
+    // TODO: validate actual contents of rows, not just count.
+    PAssert.thatSingleton(
+        output.apply("Count All", Count.<KV<String, Integer>>globally()))
+        .isEqualTo(1000L);
+
+    PAssert.that(output
+        .apply("Count Scientist", Count.<String, Integer>perKey())
+    ).satisfies(new AssertCountFn(100L));
+
+    pipeline.run().waitUntilFinish();
+  }
+
+  /**
+   * Tests writes to a postgres database.
+   *
+   * <p>Write Tests must clean up their data - in this case, it uses a new table every test run so
+   * that it won't interfere with read tests/other write tests. It uses finally to attempt to
+   * clean up data at the end of the test run.
+   * @throws SQLException
+   */
+  @Test
+  public void testWrite() throws SQLException {
+    DataSource dataSource = getDataSource();
+
+    String tableName = null;
+
+    try {
+      tableName = JdbcTestDataSet.createWriteDataTable(dataSource);
+
+      TestPipeline pipeline = TestPipeline.create();
+
+      ArrayList<KV<Integer, String>> data = new ArrayList<>();
+      for (int i = 0; i < 1000; i++) {
+        KV<Integer, String> kv = KV.of(i, "Test");
+        data.add(kv);
+      }
+      pipeline.apply(Create.of(data))
+          .apply(JdbcIO.<KV<Integer, String>>write()
+              .withDataSourceConfiguration(JdbcIO.DataSourceConfiguration.create(dataSource))
+              .withStatement(String.format("insert into %s values(?, ?)", tableName))
+              .withPreparedStatementSetter(new PutKeyInColumnOnePutValueInColumnTwo()));
+
+      pipeline.run().waitUntilFinish();
+
+      try (Connection connection = dataSource.getConnection()) {
+        try (Statement statement = connection.createStatement()) {
+          try (ResultSet resultSet = statement.executeQuery("select count(*) from "
+              + tableName)) {
+            resultSet.next();
+            int count = resultSet.getInt(1);
+
+            Assert.assertEquals(2000, count);
+          }
+        }
+      }
+      // TODO: Actually verify contents of the rows.
+    } finally {
+      JdbcTestDataSet.cleanUpDataTable(dataSource, tableName);
```

Comment:
thoughts on here vs `tearDown`? Things like unit testing timeouts, if set, would cause this code to not get run.